### ISSUE TITLE
TCP mode?

### ISSUE THREAD

**Issue by keithw** [mobile-shell/mosh#13]
Allow degenerate use of TCP and HTTP CONNECT modes for people who can't use UDP.

**Comment by pickfire** [comment]
+1, so that people could use mosh with tor.

**Comment by forkmind** [comment]
+1

**Comment by rzr** [comment]
@zedinosaur said :
> there's TCP Minion which allows datagrams over TCP 

Have you been using it with mosh ?

For the curious ones :
http://dedis.cs.yale.edu/2009/tng/papers/pfldnet10-slides.pdf

**Comment by beweisj** [comment]
+1 This is the only reason I never adopted mosh.

**Comment by mkorkalo** [comment]
+1 to mosh over SSH automatic fallback, it makes much sense to me. I also don't have a problem with ssh reconnection, it's still way better than direct SSH for the high latency use cases.

**Comment by UralZima** [comment]
+1. Very need this

**Comment by tobimensch** [comment]
I'm currently stuck with throttled 3G, and even though UDP works in principle, it looks like my provider is crippling it in every possible way. When many packages are sent in a short timeframe the data loss can be anywhere from 45% to 100%. Even though 64kbits ought to be plenty fast for the application I want to use, this leads to mosh losing contact for minutes at a time.

I'd like to create a bounty for whoever can implement a mosh over tcp mode, that is also as efficient as possible. (Doesn't really make sense when it ends up not being better/faster than regular ssh)

Is anyone interested in implementing it? What amount of money would you need to get motivated enough to actually complete this feature?

I can only chip in a little bit. Who else would donate for this?

This feature request has been open for years. It's obviously possible to solve this theoretically, which means no developer was motivated enough to do it so far. That's why I think a bounty could help to fix this.

**Comment by domo141** [comment]
@tobimensch: I've just done mosh support to my ioiomxtx tunneling and multiplexing software at

https://github.com/domo141/ioiomxtx

basically mosh datagrams multiplexed to TCP tunnel, with help of ldpreload library for port matching.

Works pretty well, and could be ported to work over other software than mxtx (e.g. ssh).

**Comment by tobimensch** [comment]
@domo141 
Sounds great. Can you write a short tutorial on how to use this with mosh?

**Comment by domo141** [comment]
@tobimensch 
I added tl;dr; section to ioiomxtx README.md which shows minimal steps to try mosh. 

I also added mxtx-mosh.md which describes how mosh(1) and mxtx-mosh(1) sets up mosh connections.

Now that I've thought it at bit it would not be too hard to modify mxtx-dgramtunneld to connect via
ssh instead of mxtx for tunneling mosh traffic is one so desired...

**Comment by jdstroy** [comment]
In case anyone else stumbles across this problem.... 

This feature will be necessary for environments that only have Internet access through a proxy server (such as a corporate SOCKS proxy, and not just going through Tor).

**Comment by fancsali** [comment]
Yes, that would be the dog's. I reckon, going through a Proxy (SOCKS or HTTP CONNECT) would be an instant hit: that would be also flexible:
- You really want Tor -- check;
- Corporate proxy -- check;
- OpenSSH has a built in SOCKS feature, so the tunnel-over-ssh --  so check!

**Comment by fancsali** [comment]
Just to add to my earlier comment: I've been thinking about this, and I reckon, we're actually looking at two separate – but connected – issues here:

1. What do do, if we have no UDP connectivity? (But *direct* TCP is fine)

I think, this is  the easier one. I've done a quick and dirty UDP-TCP forwarder in Python, that prefixes the datagram from a UDP socket with a two-byte lenght field, and pushes that through a TCP one. If the TCP side disconnects, just tries to reconnect. MOSH is more than happy to talk to the server through that. I took the liberty to peek into the code for mosh, and I reckon, something along those lines shouldn't be too hard to add.

2. What to do if even TCP is only available through some proxy?
This is a bigger problem, IMHO, as the nuances of the different proxies vary. Also, I can see a rather large problem around the problems around #285. If we cannot even do an ssh, then the whole setup becomes inconvenient. Unfortunately using `LD_PRELOAD` is not always feasible either (e.g. MacOSX will not let you much around with that and run `SSH`)

However, this second issue is easier to solve with out-of-the-box tools, I believe. Worst case I set up a tcp tunnel with socat, and make ssh connect locally:
```
$ socat tcp4-listen:60000 socks4a:<proxy-ip>:<hostname>:<port>,socksport=1080

# ~/.ssh/config
Host myserver
    HostName localhost
    Port 8022
```

... or something along those lines.

I reckon, the above broad scope is perhaps the reason, there have been no volunteers for this; even with the possibility of a bounty, as proposed by @tobimensch.

So, I'd like to propose let's limit the scope to the bare minimum, perhaps, and try to find some help like that.

**Comment by xanoni** [comment]
Could someone please update interested users (e.g., me :) on if it's now possible to use `mosh` with `sshd` behind a `tor` hidden service, and how? I'm OK with a hack, if it performs well. Thank you.

**Comment by andersk** [comment]
@xanoni That’s not possible: Mosh uses UDP connections to IP addresses, and Tor hidden services have neither UDP connections nor IP addresses. You could try routing a VPN over Tor and running Mosh over the VPN, but it’s probably going to perform worse than SSH.

**Comment by xanoni** [comment]
Oh okay. I skipped over the thread and it seemed people had found solutions (e.g., this here https://github.com/mobile-shell/mosh/issues/13#issuecomment-369731590), but I didn't check the details. Sad.

**Comment by xanoni** [comment]
For people in this thread who haven't found a solution, yet ... you might want to look at the I2P protocol instead, the C++ implementation (`i2pd`) natively supports UDP.

See here: https://i2pd.readthedocs.io/en/latest/user-guide/tunnels/#udp-tunnels

**Comment by butterfly1924** [comment]
how can we use mosh-tcp now?

**Comment by OliverOffing** [comment]
> For people in this thread who were trying to use Mosh with Tor but haven't found a solution, yet ... you might want to look at the I2P protocol instead, the C++ implementation (i2pd) natively supports UDP.

There are [instructions here on how to set up Mosh over I2P](https://stacker.news/items/31972), although the initial connection was inconsistent.

**Comment by fancsali** [comment]
Great article.

This solves the use-case of accessing your hidden machine. (Though, as I2P
seems to be much less popular, I reckon there are much fewer nodes than
Tor, hence the stability/performance might be a hit and miss, as alluded to
in the article too.)

The other use-case when I want to access (my) ordinary server(s) on public
IP from any dodgy/restricted network when out and about is still an issue.
I2P to clearnet relaying is not really a thing.

**Comment by sudo-gera** [comment]
# [Secure Socket Funneling](https://github.com/securesocketfunneling/ssf#secure-socket-funneling) is useful here.

## How to use:

Prepare 3 terminal tabs on client.
### Tab 1:
```
ssh -p REM_PORT REM_HOST -L 8011:127.0.0.1:8011 -L 2222:127.0.0.1:22 # change last 22 to actual port of remote ssh server
```
(now you're logged in on remote server)
```
cd ssf-linux-x86_64-3.0.0/ # or where is it installed
./ssfd
```
### Tab 2:
```
cd ssf-linux-x86_64-3.0.0/ # or where is it installed
./ssf 127.0.0.1 -U 3333:127.0.0.1:3333
```
### Tab 3:
```
mosh --ssh='ssh -p 2222' -p 3333 127.0.0.1
```
